File: Python/day19/AOC.py
from collections import defaultdict

def can_construct_design(design: list[str], patterns: list[str]) -> bool:
    n = len(design)
    dp = [False] * (n + 1)
    dp[0] = True
    
    patterns_by_length = defaultdict(set)
    for pattern in patterns:
        patterns_by_length[len(pattern)].add(pattern)
    
    for i in range(1, n + 1):
        for length, patterns_of_length in patterns_by_length.items():
            if i >= length:
                substring = design[i - length:i]
                if substring in patterns_of_length:
                    dp[i] = dp[i] or dp[i - length]
                    if dp[i]:  
                        break
    
    return dp[n]

# For both part 1 and two, sorting the patterns by length and storing them in a dictionary
# allows for a much faster solution, over x10 faster than the naive approach.
def part1(input: str) -> int:
    towel_patterns = input.split("\n\n")[0].split(", ")
    desired_designs = input.split("\n\n")[1].split("\n")

    return sum([1 for design in desired_designs if can_construct_design(design, towel_patterns)])

def count_arrangements(design: list[str], patterns:list[str]) -> int:
    n = len(design)
    dp = [0] * (n + 1)
    dp[0] = 1

    patterns_by_length = defaultdict(set)
    for pattern in patterns:
        patterns_by_length[len(pattern)].add(pattern)
        
    for i in range(1, n + 1):
        for length, patterns_of_length in patterns_by_length.items():
            if i >= length:
                substring = design[i - length:i]
                if substring in patterns_of_length:
                    dp[i] += dp[i - length]
                    
    return dp[n]

def part2(input: str) -> int:
    towel_patterns = input.split("\n\n")[0].split(", ")
    desired_designs = input.split("\n\n")[1].split("\n")

    return sum([count_arrangements(design, towel_patterns) for design in desired_designs])

def main():
    input = open("input.txt", "r").read()
    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

if __name__ == "__main__":
    main()

File: Go/day19/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
)

var cache = map[string]int{}

func countCombos(towel string, patterns []string) int {
	if _, ok := cache[towel]; ok {
		return cache[towel]
	}
	if len(towel) == 0 {
		return 1
	}
	count := 0
	for _, p := range patterns {
		if strings.HasPrefix(towel, p) {
			count += countCombos(towel[len(p):], patterns)
		}
	}
	cache[towel] = count
	return count
}

func part1(input string) int {
	towelPatterns := strings.Split(strings.Split(input, "\n\n")[0], ", ")
	desiredPatters := strings.Split(strings.Split(input, "\n\n")[1], "\n")
	count := 0
	for _, t := range desiredPatters {
		c := countCombos(t, towelPatterns)
		if c > 0 {
			count++
		}
	}

	return count
}

func part2(input string) int {
	towelPatterns := strings.Split(strings.Split(input, "\n\n")[0], ", ")
	desiredPatters := strings.Split(strings.Split(input, "\n\n")[1], "\n")
	count := 0
	for _, t := range desiredPatters {
		count += countCombos(t, towelPatterns)
	}

	return count
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := string(input)
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
